constructor/conda_interface.py

```diff
@@ -87,20 +91,94 @@ def get_repodata(url):
             k: v for k, v in full_repodata['packages'].items() if v['name'] in NAV_APPS}
+
+        # Minify the included repodata
         for package in used_packages:
-            for key in ('packages', 'packages.conda'):
-                if package in full_repodata.get(key, {}):
-                    used_repodata[key][package] = full_repodata[key][package]
+            key = 'packages.conda' if package.endswith(".conda") else 'packages'
+            if package in full_repodata.get(key, {}):
+                used_repodata[key][package] = full_repodata[key][package]
+                continue
+            # If we're transcoding packages, fix-up the metadata
+            if package.endswith(".conda"):
+                original_package = package[:-len(".conda")] + ".tar.bz2"
+                original_key = "packages"
+            elif package.endswith(".tar.bz2"):
+                original_package = package[:-len(".tar.bz2")] + ".conda"
+                original_key = "packages.conda"
+            else:
+                print("WARNING: package type is unknown for: %s" % package)
+                continue
+            if original_package in full_repodata.get(original_key, {}):
+                data = deepcopy(full_repodata[original_key][original_package])
+                pkg_fn = join(info["_download_dir"], package)
+                data["size"] = os.stat(pkg_fn).st_size
+                data["sha256"] = hash_files([pkg_fn], algorithm='sha256')
+                data["md5"] = hash_files([pkg_fn])
+                used_repodata[key][package] = data
+
+        write_cache_types = info.get("write_cache_types", None)
+        if not write_cache_types:
+            conda_exe_fn = os.path.basename(info["_conda_exe"])
+            if "conda.exe" in conda_exe_fn:
+                write_cache_types = ["conda"]
+            elif "micromamba" in conda_exe_fn:
+                write_cache_types = ["mamba"]
+            else:
+                write_cache_types = ["conda"]
+                print(
+                    "WARNING: Failed to detect write_cache_types using",
+                    conda_exe_fn,
+                    "defaulting to",
+                    write_cache_types,
+                )
+
+        for write_cache_type in write_cache_types:
+            if write_cache_type == "conda":
+                _write_repodata_conda(used_repodata, cache_dir, url)
+            elif write_cache_type == "mamba":
+                _write_repodata_mamba(used_repodata,cache_dir, url)
+            else:
+                raise NotImplementedError("unsupported write_cache_type: %s" % write_cache_type)
+
+
+    def _write_repodata_conda(used_repodata, cache_dir, url):
+        repodata_filename = _cache_fn_url(used_repodata['_url'].rstrip("/"))
         with open(join(cache_dir, repodata_filename), 'w') as fh:
             json.dump(used_repodata, fh, indent=2)
+
+
+    def _write_repodata_mamba(used_repodata, cache_dir, url):
+        try:
```

Comment:
do we really need this? Because it looks a bit like a bug in mamba if that's the case. We should be able to read a repodata.json without any additional metadata

Reply:
I concur that if this is a hack to fix something in mamba's implementation, this shouldn't be added here. To get a better understanding, could you elaborate why this isn't using the same repodata cache format as conda?

Reply:
@wolfv is probably in the best position to answer.

From my external perspective, assuming that there isn't any conda specification/documentation that I've missed:
* the cache format is an implementation detail that `constructor` is abusing
* either implementation could change how they cache the repodata at any time

The right solution would probably be to instead install from a local channel and let conda/mamba take care of handling that however they please (i.e. `conda.exe install -c file://$PREFIX/repodata.json ...`).

Reply:
We actually do use the same format that conda uses (which is writing a bit of metadata about the download at the beginning of the json file, namely `etag`, `timestamp` and `url` or something along those lines).

I think we maybe should ignore missing etag/... if we are in offline mode. Would that help @chrisburr ?

Reply:
Also for the filename we're supposedly using the same algorithm as conda, which is `MD5(url)[0:8]` or something along those lines.
It's possible that the URL we're using here is different for conda vs. mamba.